lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,316 @@
+  } else {
+    validCaseStmtType = true;
+  }
+
+  const auto &caseList{
+      std::get<std::list<parser::CaseConstruct::Case>>(selectCaseConstruct.t)};
+  bool defaultCaseFound{false};
+  SelectCaseHelper selectCaseStmts(context_.messages());
+
+  for (const auto &cases : caseList) {
+    const auto &caseStmt{
+        std::get<parser::Statement<parser::CaseStmt>>(cases.t)};
+    const auto &caseSelector{
+        std::get<parser::CaseSelector>(caseStmt.statement.t)};
+    if (std::holds_alternative<parser::Default>(caseSelector.u)) {
+      if (!defaultCaseFound) {
+        defaultCaseFound = true;
+      } else {  // C1146 (R1140)
+        context_.Say(caseStmt.source,
+            "Not more than one of the selectors of SELECT CASE statement may be default"_err_en_US);
+      }
+    } else if (validCaseStmtType) {
+      const auto &caseValueRangeList{
+          std::get<std::list<parser::CaseValueRange>>(caseSelector.u)};
+      for (const auto &caseValues : caseValueRangeList) {
+        if (const auto *constCase{
+                std::get_if<parser::Scalar<parser::ConstantExpr>>(
+                    &caseValues.u)}) {
+          if (!IsValidSelectCaseType(selectCaseStmtType.value(),
+                  constCase->thing, caseStmt.source)) {
+            continue;
+          }
+
+          if (selectCaseStmtType.value().category() == TypeCategory::Integer) {
+            selectCaseStmts.Insert(
+                GetIntValue(constCase->thing).value(), caseStmt.source);
+          } else if (selectCaseStmtType.value().category() ==
+              TypeCategory::Character) {
+            selectCaseStmts.Insert(
+                GetStringValue(constCase->thing).value(), caseStmt.source);
+          } else {  // TypeCategory::Logical
+            selectCaseStmts.Insert(
+                GetBoolValue(constCase->thing).value(), caseStmt.source);
+          }
+        } else {
+          if (selectCaseStmtType.value().category() ==
+              TypeCategory::Logical) {  // C1148 (R1140)
+            context_.Say(caseStmt.source,
+                "SELECT CASE expression of type LOGICAL must not have range of case value"_err_en_US);
+            continue;
+          }
+          const auto &rangeCase{
+              std::get<parser::CaseValueRange::Range>(caseValues.u)};
+          const auto &lower{rangeCase.lower};
+          if (lower &&
+              !IsValidSelectCaseType(selectCaseStmtType.value(),
+                  lower.value().thing, caseStmt.source)) {
+            continue;
+          }
+
+          const auto &upper{rangeCase.upper};
+          if (upper &&
+              !IsValidSelectCaseType(selectCaseStmtType.value(),
+                  upper.value().thing, caseStmt.source)) {
+            continue;
+          }
+
+          if (selectCaseStmtType.value().category() == TypeCategory::Integer) {
+            const auto lowerVal = lower
+                ? std::optional(GetIntValue(lower->thing).value())
```

Comment:
Why not just `GetIntValue(lower->thing)`?

Reply:
In select case, the upper range and lower range values are optional in certain cases. 
Eg: case (2:)
      case (:1)
The range pair has to be saved in the same fashion. so std::optional is being used.

Reply:
You miss the point.  `GetIntValue()` returns an `optional`.  Why not use that result directly?

Reply:
Thanks for the detailed review Klauser. I was missing the point, GetIntValue() returns an optional.